pkg/spdk/replica.go

```diff
@@ -1657,30 +1905,198 @@ func (r *Replica) doCleanupForRebuildingDst(spdkClient *spdkclient.Client, clean
+		// Otherwise, try to reuse the existing lvol
+		// TODO: For a corrupted or outdated snapshot lvol, we should create a rebuilding lvol before delete it. So that its clusters will be coalesced into the rebuilding lvol, then we can start a delta shallow copy
+		// TODO: In the future, we can move the below logic to r.RebuildingDstShallowCopyStart after the delta shallow copy API is ready
+		dstSnapBdevLvol := bdevLvolMap[dstSnapshotLvolName]
+		snaplvolSnapshotTimestamp := dstSnapBdevLvol.DriverSpecific.Lvol.Xattrs[spdkclient.SnapshotTimestamp]
+		snaplvolActualSize := dstSnapBdevLvol.DriverSpecific.Lvol.NumAllocatedClusters * defaultClusterSize
+		isIntactSnap := srcSnapSvcLvol.SnapshotTimestamp == snaplvolSnapshotTimestamp &&
+			srcSnapSvcLvol.ActualSize == snaplvolActualSize &&
+			srcSnapSvcLvol.SnapshotChecksum != "" && srcSnapSvcLvol.SnapshotChecksum == dstSnapBdevLvol.DriverSpecific.Lvol.Xattrs[spdkclient.SnapshotChecksum]
+		// For now directly delete the corrupted or outdated snapshot lvol and start a full shallow copy since we cannot validate existing data during the shallow copy
+		if !isIntactSnap {
+			for _, childLvolName := range dstSnapBdevLvol.DriverSpecific.Lvol.Clones {
+				if _, err := spdkClient.BdevLvolDetachParent(spdktypes.GetLvolAlias(r.LvsName, childLvolName)); err != nil {
+					return "", errors.Wrapf(err, "failed to decouple the child lvol %s from the corrupted or outdated snapshot lvol %s for dst replica %v rebuilding snapshot %s shallow copy prepare", childLvolName, dstSnapshotLvolName, r.Name, snapshotName)
+				}
+			}
+			if _, err := spdkClient.BdevLvolDelete(dstSnapBdevLvol.Aliases[0]); err != nil {
+				return "", errors.Wrapf(err, "failed to delete the corrupted or outdated snapshot lvol %s for dst replica %v rebuilding snapshot %s shallow copy prepare", dstSnapshotLvolName, r.Name, snapshotName)
+			}
+			r.log.Infof("Replica found a corrupted or outdated snapshot lvol %s and delete it before the shallow copy", GetReplicaRebuildingLvolName(r.Name))
+		} else {
+			snapLvolAlias := spdktypes.GetLvolAlias(r.LvsName, dstSnapshotLvolName)
+			if _, err = spdkClient.BdevLvolClone(snapLvolAlias, rebuildingLvolName); err != nil {
+				return "", errors.Wrapf(err, "failed to clone rebuilding lvol %s behinds the existing snapshot lvol %s for dst replica %v rebuilding snapshot %s shallow copy prepare", rebuildingLvolName, snapLvolAlias, r.Name, snapshotName)
+			}
+			r.log.Infof("Replica found an intact snapshot lvol %s before the shallow copy", snapLvolAlias)
+			rebuildingLvolCreated = true
+		}
+	} else { // Then check if there is an expired lvol available
+		if bdevLvolMap[dstSnapshotParentLvolName] != nil { // For non-ancestor snapshot, check if dstSnapshotParentLvol has an expired lvol as child
+			for _, childLvolName := range bdevLvolMap[dstSnapshotParentLvolName].DriverSpecific.Lvol.Clones {
+				if IsReplicaExpiredLvol(r.Name, childLvolName) {
```

Comment:
What does a expired lvol mean?

Reply:
Dst replica will rename all previous non-empty head and rebuilding lvols as expired lvols for later reusage.

You can check the workflow doc for details